### PR TITLE
feat(mesh): extracts gateway info from the ingress route

### LINKED ISSUE #12: Annotate namespaces with the Gateway for ingress traffic
For ModelMesh there is the need to expose InferenceServices. The controller managing this is [`odh-model-controller`](https://github.com/opendatahub-io/odh-model-controller/).

For the Istio integration, the `odh-model-controller` is going to create VirtualServices. A gateway needs to be specified in such VirtualServices to properly expose models. Since the `odh-project-controller` has this knowledge, it is going to be very useful to have ~~labels~~ annotations over the namespaces to set the Gateway namespace/name that can be used for exposing models.

Suggested ~~labels~~ annotations are the ones already in use in the route: `maistra.io/gateway-name` and `maistra.io/gateway-namespace`.

### PR DESCRIPTION
**Pull request by bartoszmajsak**
Fixes #12

### CODE REVIEW

File: controllers/project_mesh_controller_test.go [hunk 1]
```diff
@@ -228,23 +224,80 @@ var _ = When("Namespace is created", Label(labels.EvnTest), func() {
...
-					Name:      route.Name,
-				}, actualRoute)
+			actualTestNs := &v1.Namespace{}
+			Eventually(func() string {
+				_ = cli.Get(context.Background(), types.NamespacedName{Name: testNs.Name}, actualTestNs)
+				return actualTestNs.Annotations[controllers.AnnotationGateway]
 			}).
 				WithTimeout(timeout).
 				WithPolling(interval).
-				Should(Succeed())
```
Comment (bartoszmajsak) [inline controllers/project_mesh_controller_test.go:247]: Removed from all the tests, as this is not needed. We create route in `BeforeEach`

File: controllers/project_mesh_controller.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ import (
 const (
 	AnnotationServiceMesh = "opendatahub.io/service-mesh"
 	AnnotationGatewayHost = "opendatahub.io/service-mesh-gw-host"
+	AnnotationGateway     = "opendatahub.io/service-mesh-gateway"
```
Comment (aslakknutsen) [inline controllers/project_mesh_controller.go:22]: Should we consider just "gw" for the sake of alignment with AnnotationGatewayHost?

File: controllers/project_mesh_controller_test.go [hunk 3]
```diff
@@ -193,31 +193,27 @@ var _ = When("Namespace is created", Label(labels.EvnTest), func() {
...
-			route := &routev1.Route{
+			route = &routev1.Route{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "odh-gateway",
 					Namespace: "istio-system",
 					Labels: map[string]string{
-						"app": "odh-dashboard",
+						"app":                          "odh-dashboard",
+						"maistra.io/gateway-name":      "odh-gateway",
+						"maistra.io/gateway-namespace": "opendatahub",
```
Comment (aslakknutsen) [inline controllers/project_mesh_controller_test.go:216]: maistra?

File: controllers/namespace_updater.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+)
+
+func (r *OpenshiftServiceMeshReconciler) addGatewayAnnotations(ctx context.Context, ns *v1.Namespace) error {
+	if ns.ObjectMeta.Annotations[AnnotationGatewayHost] != "" && ns.ObjectMeta.Annotations[AnnotationGateway] != "" {
+		// If annotation is present we have nothing to do
+		return nil
+	}
+
+	routes, err := r.findIstioIngress(ctx)
```
Comment (aslakknutsen) [inline controllers/namespace_updater.go:18]: Should this Route even exist in a mesh v?

File: controllers/project_mesh_controller_test.go [hunk 5]
```diff
@@ -193,31 +193,27 @@ var _ = When("Namespace is created", Label(labels.EvnTest), func() {
...
-
-			route := &routev1.Route{
+			route = &routev1.Route{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "odh-gateway",
 					Namespace: "istio-system",
 					Labels: map[string]string{
-						"app": "odh-dashboard",
+						"app":                        "odh-dashboard",
+						controllers.LabelMaistraGw:   "odh-gateway",
```
Comment (aslakknutsen) [inline controllers/project_mesh_controller_test.go:215]: Can we rely on these? I "think" the target goes beyond ossm

File: controllers/project_mesh_controller.go (hunk 2 above)
Comment (bartoszmajsak) [reply controllers/project_mesh_controller.go:22]:
```suggestion
	AnnotationGateway     = "opendatahub.io/service-mesh-gw"
```

**Comment by israel-hdez** [comment]
It is looking good.
So, it is going to be a single annotation with the format _{namespace}/{gw-name}_, which is the same as VirtualServices expect.

**Review by israel-hdez — APPROVED** [review]

**Pull request merged**